Objective-C/Tests/URLEndpointListenerTest.m

```diff
@@ -0,0 +1,162 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+#import "ReplicatorTest.h"
+#import "CBLURLEndpointListener.h"
+#import "CBLURLEndpointListenerConfiguration.h"
+#import "CollectionUtils.h"
+
+#define kPort 4984
+#define kURL [NSURL URLWithString: $sprintf(@"ws://localhost:%d/otherdb", kPort)]
+
+API_AVAILABLE(macos(10.12), ios(10.0))
+@interface URLEndpointListenerTest : ReplicatorTest
+typedef CBLURLEndpointListenerConfiguration Config;
+typedef CBLURLEndpointListener Listener;
+
+@end
+
+@implementation URLEndpointListenerTest
+
+- (Listener*) listenAt: (uint16)port {
+    Config* config = [[Config alloc] initWithDatabase: self.otherDB port: port];
+    config.disableTLS = YES;
+    Listener* list = [[Listener alloc] initWithConfig: config];
+    
+    // start listener
+    NSError* err = nil;
+    Assert([list startWithError: &err]);
+    AssertNil(err);
+    
+    return list;
+}
+
+- (void) testPort {
+    // initialize a listener
+    Config* config = [[Config alloc] initWithDatabase: self.otherDB port: kPort];
+    config.disableTLS = YES;
+    Listener* list = [[Listener alloc] initWithConfig: config];
+    AssertEqual(list.port, 0);
+    
+    // start listener
+    NSError* err = nil;
+    Assert([list startWithError: &err]);
+    AssertNil(err);
+    AssertEqual(list.port, kPort);
+    
+    // stops
+    [list stop];
+    AssertEqual(list.port, 0);
+}
+
+- (void) testEmptyPort {
+    // initialize a listener
+    Config* config = [[Config alloc] initWithDatabase: self.otherDB port: 0];
+    config.disableTLS = YES;
+    Listener* list = [[Listener alloc] initWithConfig: config];
+    AssertEqual(list.port, 0);
+    
+    // start listener
+    NSError* err = nil;
+    Assert([list startWithError: &err]);
+    AssertNil(err);
+    Assert(list.port != 0);
+    
+    // stops
+    [list stop];
+    AssertEqual(list.port, 0);
+}
+
+- (void) testBusyPort {
+    Listener* list1 = [self listenAt: kPort];
+    
+    // initialize a listener at same port
+    Config* config = [[Config alloc] initWithDatabase: self.otherDB port: kPort];
+    config.disableTLS = YES;
+    Listener* list2 = [[Listener alloc] initWithConfig: config];
+    
+    // already in use when starting the second listener
+    [self ignoreException:^{
+        NSError* err = nil;
+        [list2 startWithError: &err];
+        AssertEqual(err.code, EADDRINUSE);
+        AssertEqual(err.domain, NSPOSIXErrorDomain);
+    }];
+    
+    // stops
+    [list1 stop];
+}
+
+// TODO: https://issues.couchbase.com/browse/CBL-948
+- (void) _testURLs {
+    Config* config = [[Config alloc] initWithDatabase: self.otherDB port: kPort];
+    config.disableTLS = YES;
```

Comment:
We also need to test without disabling TLS as well.

Reply:
should we enable TLS for all tests? or only a few to verify?

Reply:
TLS mode is the default mode so all the tests should test with TLS. Ideally, we possible we test both TLS and non-TLS but some may not need such as testPort.